src/Geography.cpp

```diff
@@ -0,0 +1,236 @@
+    if (numPoints() != rhs.numPoints()) {
+        return false;
+    }
+    for (int idx = 0; idx < numPoints(); idx += 1) {
+        if (!GeographyPoint::approximatelyEqual(getPoint(idx), rhs.getPoint(idx), epsilon)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+std::string Geography::Ring::toString() const {
+    std::stringstream stream;
+    stream << "(";
+    std::string sep("");
+    for (int idx = 0; idx < numPoints(); idx += 1) {
+        const GeographyPoint &pt = getPoint(idx);
+        stream << sep
+               << pt.getLongitude()
+               << " "
+               << pt.getLatitude();
+        sep = ", ";
+    }
+    stream << ")";
+    return stream.str();
+}
+
+void Geography::Ring::serializeTo(ByteBuffer &buffer, bool reverseit) const {
+    buffer.putInt8(0);
+    buffer.putInt32(m_points.size() - 1);
+    int start = (reverseit ? m_points.size() - 1 : 0);
+    int end   = (reverseit ? 0 : m_points.size() - 1);
+    int delta = (reverseit ? -1 : 1);
+    for (int idx = start; idx != end; idx += delta) {
+        double x, y, z;
+        m_points[idx].getXYZCoordinates(x, y, z);
+        buffer.putDouble(x);
+        buffer.putDouble(y);
+        buffer.putDouble(z);
+    }
+    char zeros[38];
+    memset(zeros, 0, sizeof(zeros));
+    buffer.put(zeros, sizeof(zeros));
+}
+
+bool Geography::approximatelyEqual(const Geography &rhs, double epsilon) const {
+    if (numRings() != rhs.numRings()) {
+        return false;
+    }
+    for (int idx = 0; idx < m_rings.size(); idx += 1) {
+        if (!getRing(idx).approximatelyEqual(rhs.getRing(idx), epsilon)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+int32_t Geography::getSerializedSize() const {
+    // Null Geography values are
+    // just 4 bytes long - just enough
+    // for the -1.
+    if (isNull()) {
+        return 4;
+    }
+    int32_t answer
+        = 4 // size
+        + 3 // 3 internal bytes
+        + 4 // Count of the number of rings
+        + 33; // Trailing overhead
+    for (RingConstIterator idx = m_rings.begin();
+         idx != m_rings.end();
+         idx++) {
+        // Each ring has 3 coordinates for each point
+        // plus 5 + 38 bytes of overhead.  Note that we
+        // will not serialize the duplicated first point.
+        answer += (idx->numPoints()-1)*3*sizeof(double) + 5 + 38;
+    }
+    return answer;
+}
+
+int32_t Geography::serializeTo(ByteBuffer &buffer) const {
+    // Remember where we start, so that we can
+    // write the size.  We keep track of how much
+    // we actually write and return it, so that the
+    // caller can check that this and getSerializedSize
+    // agree.
+    int32_t start = buffer.position();
+    if (isNull()) {
+        buffer.putInt32(-1);
+        return 4;
+    }
+
+    // Save space for the size.  We don't
+    // really know what it is now, so we
+    // well seek back to it after we are done.
+    buffer.putInt32(0);
```

Comment:
This seems like overoptimzation.  I could write

  buffer.position(buffer.position() + 4);

But it's not clear to me how this makes much difference in the absence of
any benchmarking.  The extra function call is likely to be much more expensive
than the write of 0, and it ends up being essentially the same thing.